packages/wasm/crate/src/planner.rs

```diff
@@ -1,591 +1,389 @@
+        Some(tpr::FeeMode::ManualFee(_)) => {
+            return Err(anyhow!("Manual fee mode not yet implemented").into());
         }
+    };
+
+    let source_address_index: AddressIndex = request
+        .source
+        .map(TryInto::try_into)
+        .transpose()?
+        .unwrap_or_default();
+
+    // It's possible that adding spends could increase the gas, increasing the fee
+    // amount, and so on, so we add spends iteratively.
+
+    let mut notes_by_asset_id = BTreeMap::new();
+    for required in actions
+        .balance_with_fee_estimate(&gas_prices, &fee_tier)
+        .required()
+    {
+        // Find all the notes of this asset in the source account.
+        let records = storage
+            .get_notes(NotesRequest {
+                include_spent: false,
+                asset_id: Some(required.asset_id.into()),
+                address_index: Some(source_address_index.into()),
+                amount_to_spend: None,
+            })
+            .await?;
+        notes_by_asset_id.insert(required.asset_id, records);
+    }
 
-        // Since we over-estimate the fees to be paid upfront by a fixed multiple to account
-        // for the cost of any additional `Spend` and `Output` actions necessary to pay the fee,
-        // we need to now calculate the transaction's fee again and capture the excess as change
-        // by subtracting the excess from the required value balance.
-        //
-        // Here, tx_real_fee is the minimum fee to be paid for the transaction, with no tip.
-        let mut tx_real_fee = self.gas_prices.fee(&self.plan.gas_cost());
-
-        // Since the excess fee paid will create an additional Output action, we need to
-        // account for the necessary fee for that action as well.
-        tx_real_fee += self.gas_prices.fee(&gas::output_gas_cost());
-
-        // For any remaining provided balance, add the necessary fee for collecting:
-        tx_real_fee += Amount::from(self.balance.provided().count() as u64)
-            * self.gas_prices.fee(&gas::output_gas_cost());
-
-        // Apply the fee tier to tx_real_fee so the block proposer can receive a tip:
-        tx_real_fee = Fee::from_staking_token_amount(tx_real_fee)
-            .apply_tier(self.fee_tier)
-            .amount();
-
-        assert!(
-            tx_real_fee <= self.plan.transaction_parameters.fee.amount(),
-            "tx real fee {:?} must be less than planned fee {:?}",
-            tx_real_fee,
-            self.plan.transaction_parameters.fee.amount()
-        );
-        let excess_fee_spent = self.plan.transaction_parameters.fee.amount() - tx_real_fee;
-        self.balance += Value {
-            amount: excess_fee_spent,
-            asset_id: *STAKING_TOKEN_ASSET_ID,
+    let mut iterations = 0usize;
+
+    while let Some(required) = actions
+        .balance_with_fee_estimate(&gas_prices, &fee_tier)
+        .required()
+        .next()
+    {
```

Comment:
cc @zbuc, I think this gives us a way out of the horrible over-estimate-then-adjust-then-hope method we had to do previously. I am pretty sure it would work well enough, but also that it would work perfectly if we changed the gas schedule so that the gas cost of the `Transaction` was _exactly_ the sum of the gas of its `Action`s.

Reply:
I don't think this necessarily fixes it because:

1. Right now the gas cost of the `Transaction` **is** exactly the sum of the gas of its `Action`s
2. The reason we need to do the over-estimate-then-adjust dance is because paying the fee can involve adding an unknown number of additional `Spend` actions to the transaction, depending on the available notes. So we would still have an unknown number of additional `Spend` actions (to pay fees) that we recursively have to pay fees for.